### PR TITLE
Resolve sha and branch based on workflow type

### PR DESCRIPTION
**Pull request by EeroPaukkonen**
Originally I wanted to just put in a better branch context. But this does not make much sense if sha is not properly resolved so now this PR alters both (we could split sha alterations into separate PR but I think they go well together.)

## Question marks

1.) Should we bake in this sha (/branch) detection logic into deploy-action? General arguments:
* YES - Just plug and play. People do not need to understand where to get the sha need.
* NO - This just makes the action difficult to understand and change. And the action will have quite a few edge cases e.g. using workflow that we do not support (If we go with YES we should throw error if unsupported workflow)

~~2.) Remove branch input?~~
* ~~I'm thinking yes. Generally, people should not care about it unless they give custom sha~~ I removed it 

3.) What if the sha input is given and it does not match the branch we are resolving?
* I'm thinking to set the branch to undefined and log a warning

4.) Do we even need the branch payload?
* Kind of nice context to show in Velodrome UI.
* We can avoid using branch info in workflows mostly if we want.
* I'm on the side of keeping this info just for UI and if it is good enough we can also use it in workflows.

5.) Should we error out if using a workflow trigger that we do not support
* I think yes. But not this PR.  

## Other notes

Generally, we use 4 workflow triggers to deploy
* push
* pull_request
* workflow_run - (I'm not sure if there are any good reasons for us to use workflow_run to deploy. Maybe to have deployment run in concurrency which does not do much at least right now 🤔  and you can anyway do job concurrency. Maybe there is some good reason that I'm missing.)
* deployment_status

### CODE REVIEW

**Comment by thor** [comment]
Cool! 🙌 

1. I vote for `YES`.

2. N/A. I'm guessing there were scenarios that came to mind that prompted you to remove it?

3. Something has gone wrong. While it's nice for users to avoid errors, I think users would be more bothered by the feeling of unreliable or unpredictable behaviour. If it's very likely to happen, sure -- I can see a warning working out, but if it shouldn't happen under normal operations then my vote is to raise an error (and shake our fists at the clouds).

4. One of the scenarios for using the branch payload was to distinguish between automatic deployments to production and automatic deployments to staging. That said, I suppose any successful deployment to staging wouldn't need to know the branch unless feature branches/branches in development also deploy to the staging environment. Could we get around it?

File: src/index.ts [hunk 1]
```diff
@@ -2,29 +2,95 @@ import * as core from "@actions/core";
...
+    github.context.eventName === "pull_request" ||
+    github.context.eventName === "pull_request_target"
+  ) {
+    return (github.context.payload as PullRequest).head.sha;
+  }
+
+  // 3. workflow_run
+  if (github.context.eventName === "workflow_run") {
+    return (github.context.payload as WorkflowRun).head_sha;
+  }
```
Comment (thor) [inline src/index.ts:90]: thought non-blocking: looks like a lookup-table could be useful if we add multiple scenarios

File: src/index.ts [hunk 2]
```diff
@@ -34,6 +100,10 @@ async function deploy(parentSpan: opentelemetry.Span) {
   const branch = getBranch();
   const sha = getSha();
 
+  if (!sha) {
+    throw Error(`Invalid sha ${sha}`);
```
Comment (thor) [inline src/index.ts:104]:
```suggestion
    throw Error(`Invalid SHA: ${sha}`);
```
very nitpick

File: src/index.ts [hunk 3]
```diff
@@ -2,29 +2,95 @@ import * as core from "@actions/core";
 import * as github from "@actions/github";
 import { initTracer, traceFunction } from "./tracing";
 import * as opentelemetry from "@opentelemetry/api";
+import type {
+  DeploymentStatusEvent,
+  PullRequest,
+  WorkflowRun,
+} from "@octokit/webhooks-definitions/schema";
```
Comment (thor) [inline src/index.ts:9]: oo, nice.

**Review by thor — APPROVED** [review]
LGTM

**Comment by EeroPaukkonen** [comment]
> N/A. I'm guessing there were scenarios that came to mind that prompted you to remove it?

Yea. I want to avoid people needing to think about and get confused about branch. Allowing to input branch just creates double the possibilities for confusing end results. 


> Something has gone wrong. While it's nice for users to avoid errors, I think users would be more bothered by the feeling of unreliable or unpredictable behaviour. If it's very likely to happen, sure -- I can see a warning working out, but if it shouldn't happen under normal operations then my vote is to raise an error (and shake our fists at the clouds).
EDIT: On second thought, it doesn't really make sense to check for a branch if someone provides a SHA. A SHA might have multiple branches, so we can't guarantee which one it is. 🤷 I might even skip doing any resolving if someone provides a SHA instead of a branch reference.

There are workflows that are giving sha because the automatic sha resolving was not really supporting anything else but push. So for all these we would not have branch. Not sure if this is ideal.
I'm thinking to warn for now
And consider in to get rid of those explicit and maybe get rid of sha input. 🤔 

> One of the scenarios for using the branch payload was to distinguish between automatic deployments to production and automatic deployments to staging. That said, I suppose any successful deployment to staging wouldn't need to know the branch unless feature branches/branches in development also deploy to the staging environment. Could we get around it?

If we do not want to have branch as contextual information and just want to support deployment propagation we could just have `propagate: boolean` that would be set in workflows and possibly from Velodrome UI if some app.yml flag is set....

**Comment by EeroPaukkonen** [comment]
Did a round of testing with feature branch (sha: 84f12c15b5706ea3a9af08abadd05d71d767c550, branch: eerop/test-eerop-better-branch) and one default branch check for deployment_status
* push https://github.com/kolonialno/velodrome-testing/actions/runs/4295286644/jobs/7485500323 
* pull_request https://github.com/kolonialno/velodrome-testing/actions/runs/4294066511/jobs/7482592203 
* workflow_run https://github.com/kolonialno/velodrome-testing/actions/runs/4295295196/jobs/7485505189 
* deployment_status (feature) https://github.com/kolonialno/velodrome-testing/actions/runs/4295299144/jobs/7485514652 
* deployment_status (main) https://github.com/kolonialno/velodrome-testing/actions/runs/4300541790/jobs/7496825195

Looks good. Going to merge soon unless there we get new comments.

File: src/index.ts (hunk 1 above)
Comment (EeroPaukkonen) [reply src/index.ts]: I was lazy and did not properly look at this. We might be able to make this cleaner... maybe. But that have to be separate PR.

**Pull request merged**